hs-abci-sdk/src/Tendermint/SDK/BaseApp/Block.hs

```diff
@@ -0,0 +1,94 @@
+import           Network.ABCI.Types.Messages.FieldTypes    (ConsensusParams,
+                                                            ValidatorUpdate)
+import qualified Network.ABCI.Types.Messages.Request       as Req
+import qualified Network.ABCI.Types.Messages.Response      as Resp
+import           Polysemy                                  (Embed, Members, Sem)
+import           Polysemy.Tagged                           (Tagged (..))
+import qualified Tendermint.SDK.BaseApp.Store              as Store
+import qualified Tendermint.SDK.BaseApp.Transaction.Cache  as Cache
+import           Tendermint.SDK.BaseApp.Transaction.Effect (TxEffs, runTx)
+import           Tendermint.SDK.BaseApp.Transaction.Types  (TransactionContext (..))
+import           Tendermint.SDK.Codec                      (HasCodec (..))
+import           Tendermint.SDK.Types.Effects              ((:&))
+import           Tendermint.SDK.Types.TxResult             (TxResult (_txResultEvents))
+
+
+
+data BlockContext = BlockContext TransactionContext
+
+newBlockContext :: IO BlockContext
+newBlockContext = do
+  initialCache <- newIORef Cache.emptyCache
+  gasRemaining <- newIORef 0
+  es <- newIORef []
+  pure . BlockContext $
+    TransactionContext
+      { gasRemaining
+      , txRequiresGas = False
+      , storeCache = initialCache
+      , events = es
+      }
+
+type BlockEffs = TxEffs
+
+----------------------
+
+evalBeginBlockHandler
+  :: Members [Embed IO, Tagged 'Store.Consensus Store.ReadStore, Tagged 'Store.Consensus Store.WriteStore] r
+  => Sem (BlockEffs :& r) ()
+  -> Sem r Resp.BeginBlock
+evalBeginBlockHandler action = do
+  (BlockContext txCtx)  <- liftIO newBlockContext
+  (_, txres, cache)  <- runTx (Proxy @'Store.Consensus) txCtx action
+  case cache of
+    Just c -> do
+      Cache.writeCache c
+      pure $ Resp.BeginBlock (_txResultEvents txres)
+    _ -> pure $ Resp.BeginBlock []
+
+
+defaultBeginBlocker :: Req.BeginBlock -> Sem r ()
+defaultBeginBlocker = const $ pure ()
+
+----------------------
+
+data EndBlockResult = EndBlockResult [ValidatorUpdate] (Maybe ConsensusParams)
+    deriving (Generic)
+
+instance A.ToJSON EndBlockResult
+instance A.FromJSON EndBlockResult
+instance HasCodec EndBlockResult where
+  encode = toStrict . A.encode
+  decode s = maybe (Left "EndBlockResult failure") Right (A.decodeStrict s)
+
+
+evalEndBlockHandler
+  :: Members [Embed IO, Tagged 'Store.Consensus Store.ReadStore, Tagged 'Store.Consensus Store.WriteStore] r
+  => Sem (BlockEffs :& r) EndBlockResult
+  -> Sem r Resp.EndBlock
+evalEndBlockHandler action = do
+  (BlockContext txCtx) <- liftIO newBlockContext
+  (res, txres, cache) <- runTx (Proxy @'Store.Consensus) txCtx action
+  case (res, cache) of
+    (Just (EndBlockResult updates params), Just c) -> do
+      Cache.writeCache c
+      pure $ Resp.EndBlock updates params (_txResultEvents txres)
+    _ -> pure $ Resp.EndBlock [] Nothing []
```

Comment:
It wasn't encoded into the type system, but the only way that `cache` is `Nothing` is if the tx failed (see [here](https://github.com/f-o-a-m/kepler/blob/733352cd49cda23770bd61348385b027af2e61ef/hs-abci-sdk/src/Tendermint/SDK/BaseApp/Transaction/Effect.hs#L98-L105))

It's kind of ugly, but one solution would be do do something like 

```haskell
  (BlockContext txCtx) <- liftIO newBlockContext
  (res, txres, cache) <- runTx (Proxy @'Store.Consensus) txCtx action
  case (res, cache) of
    (Just (EndBlockResult updates params), Just c) -> do
      Cache.writeCache c
      pure $ Resp.EndBlock updates params (txres ^. txResultEvents)
    (_,Nothing) -> throwError $ (txres ^. txResultAppError)
```
where `txResultAppError` comes from [here](https://github.com/f-o-a-m/kepler/blob/175c1e7b6e0c61a1389fa9a8965dd04b9260ef11/hs-abci-sdk/src/Tendermint/SDK/BaseApp/Errors.hs#L54)

you would also need to add the `Error AppError` effect to the constraints.

Reply:
:+1: